ARMv8-A MMU support

Abhishek-brcm opened the pull request:
Add MMU support for ARMv8A. As of now, 4kB translation granule is supported.
Regions to be mapped with specific attributes are required to be
at least 4kB aligned and can be provided through platform file(soc.c).
Patches have been verified on Broadcom SoC with Cortex-A72 and on qemu_cortex_a53.

Fixes #22410

Abhishek-brcm committed:
arch: arm64: Add MMU support

Add MMU support for ARMv8A. We support 4kB translation granule.
Regions to be mapped with specific attributes are required to be
at least 4kB aligned and can be provided through platform file(soc.c).

Signed-off-by: Abhishek Shah <abhishek.shah@broadcom.com>

Abhishek-brcm committed:
arch: arm64: mmu: Add zephyr execution regions

Add zephyr execution regions(text, rodata, data, noinit, bss, etc.)
with proper attributes to translation tables.
Linker script has been modified a little to align these sections to
minimum translation granule(4 kB).

With this in place, code cannot be overwritten accidently as it is
marked read only. Similarly, execution is prohibited from data/RW
section as it is marked execute-never.

Signed-off-by: Abhishek Shah <abhishek.shah@broadcom.com>

Abhishek-brcm committed:
soc: arm64: qemu-cortex-a53: Add MMU support

Add qemu-cortex-a53 memory regions with proper attributes
to translation tables. Minimal regions to execute "hello_world"
are added as of now, More granular memory regions should be
added later as per the requirement.

Signed-off-by: Abhishek Shah <abhishek.shah@broadcom.com>

carlocaione commented on arch/arm/core/aarch64/arm_mmu.c:
```diff
@@ -0,0 +1,464 @@
+	MMU_DEBUG("enabling dcache/mmu\n");
+
+	/* currently only EL1 is supported */
+	enable_mmu_el1(flags);
```
We are currently switching to EL1 automatically at boot when `CONFIG_SWITCH_TO_EL1` is set. This is the default case since `CONFIG_SWITCH_TO_EL1` defaults to `=y` but this also means that the MMU is broken if someone sets `CONFIG_SWITCH_TO_EL1=n` and for example we are running in EL3.

Can you add a check before enabling the MMU to check whether we actually are in EL1 and gracefully print an error message otherwise?

Abhishek-brcm replied:
@carlocaione Thanks for the review.
Current MMU code does not support EL3. I will add print message and gracefully exit if not in EL1.

On a side note, we should be able to support MMU in EL3 by programming EL3 system registers equivalent of EL1 system registers(TTBR_ELx, TCR_ELx, MAIR_ELx, SCTLR_ELx), however, I have not verified it yet.

carlocaione replied:
If you can support EL3 even better. Your call.

sandeepbrcm replied:
@carlocaione , Like i cited before in your PR, I feel we can add *Zephyr in EL3*  only when there is a real need as this will have maintenance overhead to consider behavior in multiple exception levels and will add code under #ifdefs / if (currentEL == Elx). It will require everyone to run tests both EL3/EL1 support variants for any arch port changes. 
So I feel El3 support can be deferred.

carlocaione replied:
>  So I feel El3 support can be deferred.

ACK

carlocaione commented on arch/arm/core/aarch64/arm_mmu.c:
```diff
@@ -0,0 +1,464 @@
+	}
+
+	if (el == 1) {
+		tcr = TCR_EL1_RSVD | (tcr_ps_bits << TCR_EL1_IPS_SHIFT);
```
What `TCR_EL1_RSVD` is for? Is this implementation defined?

Abhishek-brcm replied:
will look into this and get back.

Abhishek-brcm replied:
In `TCR_EL1`, bit `[15:0]` is for memory region addressed `TTBR0_EL1 `and bit `[31:16]` is for memory region addressed `TTBR1_EL1`.

TCR_EL1_RSVD  is `(1 << 30)`. It used to indicate translation granule along with bit 30 (`TG1[31:30]`). It is defined for some cores and reserved for some. So, yes, this field is implementation defined.

However, we have disabled translation table walk using `TTBR1_EL1`, so this field should not matter run time.

Reference: ARMv8-A ARM (DDI 0487E.a): D13.2.112 `TCR_EL1`, Translation Control Register (EL1)

carlocaione replied:
Yes, this is what I thought. Being something custom (implementation defined) I would probably leave that out.

Abhishek-brcm replied:
Not touching implementation defined bits in TCR.

carlocaione commented on arch/arm/core/aarch64/arm_mmu.c:
```diff
@@ -0,0 +1,464 @@
+ * (va_bits <= 21)	 - base level 3
+ * (22 <= va_bits <= 30) - base level 2
+ * (31 <= va_bits <= 39) - base level 1
+ * (40 <= va_bits <= 47) - base level 0
```
s/CONGIG_ARM64_VA_BITS/CONFIG_ARM64_VA_BITS/ and s/47/48/

carlocaione commented on arch/arm/core/aarch64/arm_mmu.c:
```diff
@@ -0,0 +1,464 @@
+
+/* Virtual Address Index within given translation table level */
+#define XLAT_TABLE_VA_IDX(va_addr, level) \
+	((va_addr >> LEVEL_TO_VA_SIZE_SHIFT(level)) & 0x1ffULL)
```
Instead of hardcoding `0x1ffULL` what about `XLAT_TABLE_ENTRIES - 1`? If you are going to change the size of the translation table it's better to just change one parameter.

Abhishek-brcm replied:
Will use `XLAT_TABLE_ENTRIES - 1`.

carlocaione commented on arch/arm/core/aarch64/arm_mmu.c:
```diff
@@ -0,0 +1,464 @@
+		__aligned(NUM_BASE_LEVEL_ENTRIES * sizeof(u64_t));
+
+static u64_t xlat_tables[CONFIG_MAX_XLAT_TABLES][XLAT_TABLE_ENTRIES]
+		__aligned(XLAT_TABLE_SIZE);
```
Can we use `XLAT_TABLE_ENTRIES * sizeof(u64_t)` for the sake of consistency?

Abhishek-brcm replied:
I didn't get this. 

`CONFIG_MAX_XLAT_TABLES ` is required number of translation tables, each with `XLAT_TABLE_ENTRIES`(i.e. 512 entries), and each entry of 8 B, so `CONFIG_MAX_XLAT_TABLES  `with each translation table of (8* 512 = 4kB).

`__aligned(XLAT_TABLE_SIZE)` is used because `xlat_tables `have to be `XLAT_TABLE_SIZE`(4kB in our case) aligned.

carlocaione replied:
my suggestion was only about:
```
__aligned(XLAT_TABLES_ENTRIES * sizeof(u64_t));
```

Abhishek-brcm replied:
ok, got it.

carlocaione commented on soc/arm/qemu_cortex_a53/mmu_regions.c:
```diff
@@ -0,0 +1,28 @@
+static const struct arm_mmu_region mmu_regions[] = {
+
+	MMU_REGION_FLAT_ENTRY("GIC",
+			      0x8000000, 0x20000,
```
Can we use `DT_INST_0_ARM_*` (and `DT_INST_0_ARM_PL011_*` for the UART) instead of hardcoding memory address and size?

Abhishek-brcm replied:
yes, will change.

carlescufi commented:
@Abhishek-brcm mind shortening the lines in your commit message as shown [here](https://github.com/zephyrproject-rtos/zephyr/pull/22446#issuecomment-581516919)?

carlocaione commented on arch/arm/core/aarch64/arm_mmu.c:
```diff
@@ -0,0 +1,464 @@
+#define XLAT_TABLE_SIZE		(1U << XLAT_TABLE_SIZE_SHIFT)
+
+#define XLAT_TABLE_ENTRY_SIZE_SHIFT	3U /* Each table entry is 8 bytes */
+#define XLAT_TABLE_LEVEL_MAX	3U
```
I'm a bit confused about the relation between `XLAT_TABLE_LEVEL_MAX` and `CONFIG_MAX_XLAT_TABLES`. Aren't those supposed to be the same value?

Abhishek-brcm replied:
`XLAT_TABLE_LEVEL_MAX `indicates maximum level of translation table in ARMv8-A, i.e. 3.
And `CONFIG_MAX_XLAT_TABLES `indicates total number of translation tables(minus base translation table).

So, `XLAT_TABLE_LEVEL_MAX `will remain same for all platforms.
However, `CONFIG_MAX_XLAT_TABLES `can change depending memory(normal+device) map of the SoC. That's why it is config option.

carlocaione commented on arch/arm/core/aarch64/arm_mmu.c:
```diff
@@ -0,0 +1,464 @@
+#if MMU_DEBUG_PRINTS
+/* To dump page table entries while filling them, set DUMP_PTE macro */
+#define DUMP_PTE		0
+#define MMU_DEBUG(fmt, ...)	printk(fmt, ##__VA_ARGS__)
```
Any reason to not use the `LOG_*` macros?

Abhishek-brcm replied:
`MMU_DEBUG `macro is introduced to debug the MMU code itself. For example, to dump page table entries.
I can probably change few prints to use `LOG_ *`macro if not specific to MMU debug.

carlocaione commented on arch/arm/core/aarch64/Kconfig:
```diff
@@ -90,6 +90,95 @@ config GEN_ISR_TABLES
+if ARM_MMU
+
+choice
+	prompt "Page size"
```
Do we need this if you are hardcoding `PAGE_SIZE` directly into the MMU code?

Abhishek-brcm replied:
yes right, I will just remove this.

carlocaione commented:
LGTM now. Thanks.

sandeepbrcm commented:
+1

timork commented on arch/arm/core/aarch64/arm_mmu.c:
```diff
@@ -0,0 +1,476 @@
+	/* AP bits for Data access permission */
+	desc |= (attrs & MT_RW) ? PTE_BLOCK_DESC_AP_RW : PTE_BLOCK_DESC_AP_RO;
+
+	/* the access flag */
```
The “The AP[2:1] data access permissions bits” must enable using MMU from EL0 also. 

#define PTE_BLOCK_DESC_EL0_EN                          (1UL << 6)
/* AP bits for EL0/El1 access permission */
desc |= PTE_BLOCK_DESC_EL0_EN;

Abhishek-brcm replied:
Normally, we would be executing user applications at EL0. Currently, I do not see any EL0 software getting executed with zephyr. So, I am not sure if this change should be done.

timork replied:
@carlocaione What the execution level for user task in current aarch64 implementation?

carlocaione replied:
Everything is run at EL3

timork reviewed (changes requested):
Support for NEW LIB: Currently initialization of MMU runs by using “SYS_INIT(arm_mmu_init, PRE_KERNEL_1,….”. 
This initialization is invoked after zeroing of bss and copy of data sections. For correct NEW LIB support we need MMU working before care of bss and data sections.

stephanosio commented on include/arch/arm/aarch64/arm_mmu.h:
```diff
@@ -0,0 +1,166 @@
+#define PTE_BLOCK_DESC_AF		(1UL << 10)
+#define PTE_BLOCK_DESC_NG		(1UL << 11)
+#define PTE_BLOCK_DESC_PXN		(1UL << 53)
+#define PTE_BLOCK_DESC_UXN		(1UL << 54)
```
Nit: `ULL` would be a better choice (and aligns with what `TCR` definitions are doing).

stephanosio commented on include/arch/arm/aarch64/scripts/linker.ld:
```diff
@@ -148,7 +151,6 @@ SECTIONS
 
     SECTION_PROLOGUE(_TEXT_SECTION_NAME_2,,)
```
The recent Cortex-M `_TEXT_SECTION_NAME_2` clean-up could be done here too? (or possibly in a separate PR)

stephanosio replied:
Done, see #22673.

stephanosio commented:
> Support for NEW LIB: Currently initialization of MMU runs by using “SYS_INIT(arm_mmu_init, PRE_KERNEL_1,….”.
> This initialization is invoked after zeroing of bss and copy of data sections. For correct NEW LIB support we need MMU working before care of bss and data sections.

@timork This should only be a problem when the section start or end address is unaligned (newlib `memcpy` and `memset` functions have no reason to perform unaligned accesses when the memory addresses are aligned).

One exception is if the newlib nano variant is used. The nano `memcpy` implementation, for instance, always produces unaligned memory accesses:
https://github.com/bminor/newlib/blob/d8ccbcdaccff12ca9509358d31114515acd20590/newlib/libc/string/memcpy.c#L53-L64

Noting that there is absolutely no reason to use newlib nano on AArch64, it might make the most sense to simply disallow `CONFIG_NEWLIB_LIBC_NANO=y` when `CONFIG_ARM64=y`.

timork commented:
@stephanosio 
On aarch64 asm optimized memory operations are used:
https://github.com/bminor/newlib/blob/d8ccbcdaccff12ca9509358d31114515acd20590/newlib/libc/machine/aarch64/memcpy.S

The assumptions on this impl. is unaligned accesses machine suporrt.

stephanosio commented:
> @stephanosio
> On aarch64 asm optimized memory operations are used:
> https://github.com/bminor/newlib/blob/d8ccbcdaccff12ca9509358d31114515acd20590/newlib/libc/machine/aarch64/memcpy.S

@timork For newlib nano, even on AArch64, the dumb never-aligned byte-level copy implementation is used:
https://github.com/bminor/newlib/blob/d8ccbcdaccff12ca9509358d31114515acd20590/newlib/libc/machine/aarch64/memcpy.S#L61-L63

> The assumptions on this impl. is unaligned accesses machine suporrt.

Unaligned accesses generate fault when the MMU is disabled, which I thought was the rationale behind why you said "For correct NEW LIB support we need MMU working before care of bss and data sections."

timork commented:
@stephanosio 
> Unaligned accesses generate fault when the MMU is disabled, which I thought was the rationale behind why you said "For correct NEW LIB support we need MMU working before care of bss and data sections."

Yes

>Noting that there is absolutely no reason to use newlib nano on AArch64, it might make the most sense to simply disallow CONFIG_NEWLIB_LIBC_NANO=y when CONFIG_ARM64=y.

Except cases when perf. of memory operations routines is important.

carlocaione commented:
> Noting that there is absolutely no reason to use newlib nano on AArch64, it might make the most sense to simply disallow `CONFIG_NEWLIB_LIBC_NANO=y` when `CONFIG_ARM64=y`.

We are already doing that with https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/toolchain/zephyr/0.11/Kconfig#L8

stephanosio commented:
> > Noting that there is absolutely no reason to use newlib nano on AArch64, it might make the most sense to simply disallow `CONFIG_NEWLIB_LIBC_NANO=y` when `CONFIG_ARM64=y`.
> 
> We are already doing that with https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/toolchain/zephyr/0.11/Kconfig#L8

@carlocaione In that case, using newlib should not be a problem as long as the section addresses are aligned.

Abhishek-brcm commented on arch/arm/core/aarch64/arm_mmu.c:
```diff
@@ -0,0 +1,476 @@
+ */
+static int arm_mmu_init(struct device *arg)
+{
+	unsigned int val, flags = 0;
```
Right, will change.

Abhishek-brcm commented on arch/arm/core/aarch64/arm_mmu.c:
```diff
@@ -0,0 +1,476 @@
+	if (GET_EL(val) != MODE_EL1) {
+		LOG_WRN("Exception level not EL1, MMU not enabled!\n");
+		return -EINVAL;
+	}
```
yes, I think so, because we may face alignment issues later if we just go ahead. Will change.

Abhishek-brcm commented on arch/arm/core/aarch64/arm_mmu.c:
```diff
@@ -0,0 +1,476 @@
+	/* Ensure that MMU is already not enabled */
+	__asm__ volatile("mrs %0, sctlr_el1" : "=r" (val));
+	LOG_DBG("sctlr_el1: %x\n", val);
+	__ASSERT((val | SCTLR_M_BIT) != 0, "MMU is already enabled\n");
```
Prior stage image(e.g. bootloader) might have left MMU enabled before jumping to zephyr execution. So, I think it is safe to keep the check.

Abhishek-brcm commented on arch/arm/core/aarch64/arm_mmu.c:
```diff
@@ -0,0 +1,476 @@
+			: "memory", "cc");
+
+	/* Ensure the MMU enable takes effect immediately */
+	__ISB();
```
Nice catch! Will add isb before enabling MMU.

stephanosio commented on arch/arm/core/aarch64/arm_mmu.c:
```diff
@@ -0,0 +1,477 @@
+
+	/* Ensure that MMU is already not enabled */
+	__asm__ volatile("mrs %0, sctlr_el1" : "=r" (val));
+	LOG_DBG("sctlr_el1: %x\n", val);
```
`LOG_DBG("sctlr_el1: %llx\n", val);`

stephanosio replied:
https://gist.githubusercontent.com/stephanosio/4e1c2bd72652c4ae3f806f82253ee94d/raw/7db91923b0a402309e176dd007a8cfcf0deac366/zephyrproject-rtos-zephyr-63051.1.log

Abhishek-brcm replied:
Thanks for pointing out. Fixed.

stephanosio commented:
@ehudmarvell Could you check if the latest implementation works on your H/W platform? (especially when newlib is enabled)

tyrantlizard77 reviewed (approved):
no objections to this in the overall picture of Zephyr memory management.
ongoing discussion on the ARM particulars (which I am not deeply familiar with yet) should continue.

Abhishek-brcm committed:
arch: arm: aarch64: linker.ld: move vector table to text segment

Following changes are done:
- The vector table should be placed in text segment.
- Removed Vector relay table related entries as it is
  only applicable to aarch32.
- irq_vector_table contains ISR pointers - should be placed
  in rodata segment.
- put openocd_dbg in rodata and skip adding <linker_relocate.ld>
  as CONFIG_CODE_DATA_RELOCATION is not defined for aarch64
  currently (add later if needed).

Fixes: #22673
Signed-off-by: Abhishek Shah <abhishek.shah@broadcom.com>

carlocaione commented:
The failure on `tests/application_development/libcxx` is really suspicious since there is no clear indication on why it is failing and it seems that they are timing out without even starting QEMU.

Abhishek-brcm commented:
> The failure on `tests/application_development/libcxx` is really suspicious since there is no clear indication on why it is failing and it seems that they are timing out without even starting QEMU.

I have verified qemu_cortex_a53 boot. It would be great if someone with knowledge on Shippable test/sanitycheck can help resolve this.

carlocaione commented:
I think that the tests are failing because they are hitting https://github.com/zephyrproject-rtos/zephyr/pull/22446/commits/60b65208307792afa61193af0b3bff8b748a6bc6#diff-2d71619e0283603f202f7ce63309c0cfR302

Abhishek-brcm commented:
> I think that the tests are failing because they are hitting [60b6520#diff-2d71619e0283603f202f7ce63309c0cfR302](https://github.com/zephyrproject-rtos/zephyr/commit/60b65208307792afa61193af0b3bff8b748a6bc6#diff-2d71619e0283603f202f7ce63309c0cfR302)
> 
> ```
> #0  arch_system_halt (reason=reason@entry=4) at /home/carlo/work/proj/zephyr/zephyrproject/zephyr/kernel/fatal.c:31
> #1  0x0000000040013764 in k_sys_fatal_error_handler (reason=reason@entry=4, esf=esf@entry=0x0) at /home/carlo/work/proj/zephyr/zephyrproject/zephyr/kernel/fatal.c:43
> #2  0x00000000400137f4 in z_fatal_error (reason=reason@entry=4, esf=esf@entry=0x0) at /home/carlo/work/proj/zephyr/zephyrproject/zephyr/kernel/fatal.c:123
> #3  0x0000000040011814 in assert_post_action (file=file@entry=0x40024404 "ZEPHYR_BASE/arch/arm/core/aarch64/arm_mmu.c", line=line@entry=302)
>     at /home/carlo/work/proj/zephyr/zephyrproject/zephyr/lib/os/assert.c:45
> #4  0x00000000400128e0 in init_xlat_tables (region=region@entry=0x400221a0 <mmu_zephyr_regions+40>) at /home/carlo/work/proj/zephyr/zephyrproject/zephyr/arch/arm/core/aarch64/arm_mmu.c:302
> #5  0x0000000040012cb4 in setup_page_tables () at /home/carlo/work/proj/zephyr/zephyrproject/zephyr/arch/arm/core/aarch64/arm_mmu.c:392
> #6  arm_mmu_init (arg=<optimized out>) at /home/carlo/work/proj/zephyr/zephyrproject/zephyr/arch/arm/core/aarch64/arm_mmu.c:461
> #7  0x0000000040013680 in z_sys_device_do_config_level (level=level@entry=0) at /home/carlo/work/proj/zephyr/zephyrproject/zephyr/kernel/device.c:54
> #8  0x00000000400138fc in z_cstart () at /home/carlo/work/proj/zephyr/zephyrproject/zephyr/kernel/init.c:523
> #9  0x0000000040012650 in z_arm64_prep_c () at /home/carlo/work/proj/zephyr/zephyrproject/zephyr/arch/arm/core/aarch64/prep_c.c:31
> #10 0x000000004000ec38 in __start () at /home/carlo/work/proj/zephyr/zephyrproject/zephyr/arch/arm/core/aarch64/reset.S:145
> ```

This helps. I see that `#include <linker/cplusplus-rom.ld>` is coming in effect for `tests/application_development/libcxx` test and making image_rodata_end 4k unaligned causing MMU assert.

Abhishek-brcm committed:
arch: arm: aarch64: linker: Discard plt and got sections

plt and got sections are used for dynamic linking which
is not supported in Zephyr.

Reference: #11953
commit 3ba7097e7326 ("linker: add orphan sections to linker script")

Signed-off-by: Abhishek Shah <abhishek.shah@broadcom.com>

carlescufi commented:
@Abhishek-brcm @carlocaione there are still some tests failing CI, on the `qemu_cortex_a53` platform:
https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/63654/5/console

Abhishek-brcm commented:
> @Abhishek-brcm @carlocaione there are still some tests failing CI, on the `qemu_cortex_a53` platform:
> https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/63654/5/console

I see that I am not able to use "`LOG_`" in this driver, which is causing `tests/subsys/logging/log_core/` to fail. I tried using "`LOG_`" macro in timer driver(`drivers/timer/arm_arch_timer.c`) instead, and same failure is seen.[Also I see no timer driver using `LOG_*`] So I think that there is not issue with driver itself, but rather related to logging use. Currently, I am submitting MMU driver after removing `LOG_*`. That should make CI pass.

carlescufi commented:
@Abhishek-brcm ignore the Documentation and License checks showing in progress, those are now handled by GitHub actions. This PR is ready for merging.

Abhishek-brcm commented:
> @Abhishek-brcm ignore the Documentation and License checks showing in progress, those are now handled by GitHub actions. This PR is ready for merging.

Ok, sounds good.

ioannisg commented on arch/arm/core/aarch64/CMakeLists.txt:
```diff
@@ -20,3 +20,4 @@ zephyr_library_sources(
 
 zephyr_library_sources_ifdef(CONFIG_GEN_SW_ISR_TABLE isr_wrapper.S)
 zephyr_library_sources_ifdef(CONFIG_IRQ_OFFLOAD irq_offload.c)
+zephyr_library_sources_ifdef(CONFIG_ARM_MMU arm_mmu.c)
```
Do you anticipate the need of a sub-directory here instead of a single MMU driver source file?

Abhishek-brcm replied:
I do not anticipate need of sub-directory as of now; Especially, no vendor specific mmu code(I see some for cortex-m). We can always add sub-directory later if need arises.

ioannisg replied:
ok

ioannisg commented on arch/arm/core/aarch64/Kconfig:
```diff
@@ -90,6 +90,83 @@ config GEN_ISR_TABLES
+config ARM_MMU
+	bool "ARM MMU Support"
+	help
+	  Memory Management Unit support.
```
What are the (reasonable) dependencies here?
E.g. for Cortex-M the MPU is an optional Cortex peripheral, so the dependency is CPU_HAS_MPU.
What is the case for Cortex-A, is MMU a mandatory feature?

Also, does it make sense to build the core without MMU support if this is present? If not, this could be default y.

sandeepbrcm replied:
>>Does it make sense to build the core without MMU support if this is present ?

Cortex-A (v8/v7) MMU Hardware is mandatory feature. Sometimes for debug purpose it might make sense to build the code without MMU even if support is present.

stephanosio replied:
> Also, does it make sense to build the core without MMU support if this is present? If not, this could be default y.

It should definitely be `default y`. Unaligned accesses are not supported (RAM is treated as Device-nGnRnE) when the MMU is disabled and this means you will be seeing lots of alignment faults in real hardware.

Abhishek-brcm replied:
We will enable by default. If required for debug purpose, one can manually modify `Kconfig` and disable MMU.
Note that till now we were adding `select ARM_MMU` in board config(ex.: `soc/arm/qemu_cortex_a53/Kconfig.soc`), will remove that.

ioannisg replied:
ok good

ioannisg commented on arch/arm/core/aarch64/Kconfig:
```diff
@@ -90,6 +90,83 @@ config GEN_ISR_TABLES
+	int "Maximum numbers of translation tables"
+	default 7
+	help
+	  This option specifies maximum numbers of translation tables excluding
```
style: the maximum number of translation tables

ioannisg commented on arch/arm/core/aarch64/Kconfig:
```diff
@@ -90,6 +90,83 @@ config GEN_ISR_TABLES
+	help
+	  This option specifies maximum numbers of translation tables excluding
+	  base translation table. Based on this, translation tables are
+	  allocated compile time and used runtime as needed. If the runtime need
```
style: at* compile time, used at* runtime

ioannisg commented on arch/arm/core/aarch64/Kconfig:
```diff
@@ -90,6 +90,83 @@ config GEN_ISR_TABLES
+	  allocated compile time and used runtime as needed. If the runtime need
+	  exceeds preallocated numbers of translation tables, it will result
+	  in assert. Number of translation tables required is decided based on
+	  how many descrete memory regions (both normal and device memory)
```
discrete*

ioannisg commented on arch/arm/core/aarch64/Kconfig:
```diff
@@ -90,6 +90,83 @@ config GEN_ISR_TABLES
+if ARM_MMU
+
+config MAX_XLAT_TABLES
+	int "Maximum numbers of translation tables"
```
Do we need to add a range property to the Kconfig symbol?

Abhishek-brcm replied:
I don't think so. User can specify number of numbers of translation tables per requirement.

ioannisg commented on include/arch/arm/aarch64/arm_mmu.h:
```diff
@@ -0,0 +1,166 @@
+	/* Region Base Virtual Address */
+	u64_t base_va;
+	/* Region size */
+	u64_t size;
```
does it make sense to have this as size_t?

stephanosio replied:
Had a similar thought earlier. IMO, since this is a highly arch-specific (in particular, AArch64-specific) construct and portability is not of concern, I do not think there would be any benefit in making it `size_t`.

stephanosio replied:
Though, it would be a different story if we decide to use `struct arm_mmu_region` for AArch32 as well, but that would be in the future.

Abhishek-brcm replied:
Keeping it `u64_t` for now.

ioannisg commented on include/arch/arm/aarch64/scripts/linker.ld:
```diff
@@ -39,6 +39,8 @@
   #define ROM_ADDR (CONFIG_FLASH_BASE_ADDRESS + CONFIG_FLASH_LOAD_OFFSET)
 #endif
 
+#define PAGE_SIZE	0x1000
```
Can you document this value? Is this arbitrary selection?

stephanosio replied:
While AArch64 supports multiple page sizes, this PR only supports 4KiB (which is quite reasonable; e.g. x86 uses 4KiB as default as well).

ioannisg replied:
OK. Still, needs inline explanation

Abhishek-brcm replied:
Will add inline explanation.

ioannisg commented on include/arch/arm/aarch64/scripts/linker.ld:
```diff
@@ -114,39 +106,27 @@ SECTIONS
     _image_rom_start = ROM_ADDR;
 
-    SECTION_PROLOGUE(rom_start,,)
+    SECTION_PROLOGUE(_TEXT_SECTION_NAME,,)
```
Is it possible for this to look like Cortex-M and Cortex-R? 
rom_start can include vector tables (At least) and other applicable section, and text can include all other stuff that belong to text

sandeepbrcm replied:
> rom_start can include vector tables (At least) 

Not sure about all other sections for similarity but  vector table in particular for cortex-A / Cortex-R belong to .text  unlike Cortex-M. So it differs there.

ioannisg replied:
For Cortex-R, vector table section is in rom_start, please, check.
Why can't you do the same for Cortex-A?

Abhishek-brcm replied:
For ARMv8-A, vector table is added in `.exc_vector_table` section, and on interrupt/exception core executes from there, so we have to put it in text segment.

ioannisg replied:
Agreed; can you just do the same thing as Cortex-M and Cortex-R? vector section is inside _exc_vector_table

ioannisg replied:
@Abhishek-brcm see the vector_table.ld in arm/aarrch32

Abhishek-brcm replied:
@ioannisg currently, I have moved around these 4 sections among text/rodata segment because of MMU requirement, namely, `.exc_vector_table`, `IRQ_VECTOR_TABLE`, `.openocd_dbg` and `<linker_relocate.ld>`.  There 4 are sections read only. However, from MMU point of view, these read only sections are either executable or never-execute. We have to classify accordingly.

Some background on MMU implementation:
MMU will mark "text" segment `[_image_text_start, _image_text_end]` as code, i.e. read only and executable.
And it will mark "rodata" segment `[_image_rodata_start, _image_rodata_end]` as rodata, i.e. read only and never execute.

Justification for change:
When you suggest to make aarch64 linker file same as cortex-M or cortex-R, you are suggesting to put vector table in "read only" segment. From MMU point of view, "read only" segment is not executable, so we must put vector table in "text" segment. 

I can move other 3 sections to "rodata" segment- based on the assumption that all 3 are actually read only and never-execute. [currently, I have put only `IRQ_VECTOR_TABLE` in rodata, I can move remaining 2 also to rodata]

For cortex-R, MMU has not yet come into the picture. Once does, I think we may need similar changes there as well.

ioannisg replied:
No, i am not suggesting to put vectors in rodata; I suggested to follow the same structure as Cortex-R where .vectors are inside exc_vector_table section, and _that_ one is grouped inside rom_start.

But now I realize your point; MMU has more strict requirement regarding attributing segments, than Cortex-M; in Cortex-M we play around with the section grouping freely, because we do not define MPU sections in the flash (we only do it in SRAM).

In that case, yes, I'd rather see, then, everything that is just read-only (no exec) inside ro-data. You could even remove the sections if they are not needed (They could be added later if this becomes needed).

Anyway, this was just a point of organizing the linker script in a consistent way, I guess we can live with the current approach

Abhishek-brcm replied:
I think we are on the same page now. I will put `IRQ_VECTOR_TABLE` in rodata, `openocd_dbg` in rodata (I see some data constructs being put here by `subsys/debug/openocd.c`) and will skip adding `<linker_relocate.ld>` as `CONFIG_CODE_DATA_RELOCATION` is not defined for aarch64 currently(add later if needed).

ioannisg reviewed (changes requested):
Looks mostly OK, I added some comments and some requests for changes. I'm mostly concerned about the linker script - I'd like this to look similar to the other ARM variants, and please check the comments about the MMU Kconfig symbol
